cmd/otel-allocator/allocation/allocator.go

```diff
@@ -47,103 +58,96 @@ func (t TargetItem) hash() string {
 
 // Create a struct that holds collector - and jobs for that collector
 // This struct will be parsed into endpoint with collector and jobs info
-
+// This struct can be extended with information like annotations and labels in the future
 type collector struct {
 	Name       string
 	NumTargets int
 }
 
+type State struct {
+	// collectors is a map from a collector's name to a collector instance
+	collectors map[string]collector
+	// targetItems is a map from a target item's hash to the target items allocated state
+	targetItems map[string]TargetItem
+}
+
+func NewState(collectors map[string]collector, targetItems map[string]TargetItem) State {
+	return State{collectors: collectors, targetItems: targetItems}
+}
+
+type changes[T any] struct {
+	additions map[string]T
+	removals  map[string]T
+}
+
+func diff[T any](current, new map[string]T) changes[T] {
+	additions := map[string]T{}
+	removals := map[string]T{}
+	// Used as a set to check for removed items
+	newMembership := map[string]bool{}
+	for key, value := range new {
+		if _, found := current[key]; !found {
+			additions[key] = value
+		}
+		newMembership[key] = true
+	}
+	for key, value := range current {
+		if _, found := newMembership[key]; !found {
+			removals[key] = value
+		}
+	}
+	return changes[T]{
+		additions: additions,
+		removals:  removals,
+	}
+}
+
+type AllocatorStrategy interface {
+	Allocate(currentState, newState State) State
+}
+
+func NewStrategy(kind string) (AllocatorStrategy, error) {
+	if kind == "least-weighted" {
+		return LeastWeightedStrategy{}, nil
+	}
```

Comment:
How about inverting this and making a registry.  Strategies can register themselves in an `init()` function and this function should just be a map lookup.

```go
type AllocatorProvider func() AllocatorStrategy

var registry map[string]AllocatorProvider

func NewStrategy(name string) (AllocatorStrategy, error) {
  if p, ok := registry[name]; ok {
    return p(), nil
  }
  return nil, errors.New(...)
}

func RegisterStrategy(name string, provider AllocatorProvider) error {
  if _, ok := registry[name]; ok {
    return errors.New("already registered")
  }
  registry[name] = provider
  return nil
}
```